fem/qinterp/grad.hpp

```diff
@@ -135,34 +164,70 @@ static void Derivatives2D(const int NE,
          {
             MFEM_FOREACH_THREAD(qx,x,Q1D)
             {
-               double u = 0.0;
-               double v = 0.0;
+               double du[3] = {0.0, 0.0, 0.0};
                for (int dy = 0; dy < D1D; ++dy)
                {
-                  u += DQ1(dy,qx) * B(dy,qy);
-                  v += DQ0(dy,qx) * G(dy,qy);
+                  du[0] += DQ1(dy,qx) * B(dy,qy);
+                  du[1] += DQ0(dy,qx) * G(dy,qy);
                }
                if (GRAD_PHYS)
                {
-                  double Jloc[4], Jinv[4];
-                  Jloc[0] = j(qx,qy,0,0,e);
-                  Jloc[1] = j(qx,qy,1,0,e);
-                  Jloc[2] = j(qx,qy,0,1,e);
-                  Jloc[3] = j(qx,qy,1,1,e);
-                  kernels::CalcInverse<2>(Jloc, Jinv);
-                  const double U = Jinv[0]*u + Jinv[1]*v;
-                  const double V = Jinv[2]*u + Jinv[3]*v;
-                  u = U; v = V;
-               }
-               if (Q_LAYOUT == QVectorLayout::byVDIM)
-               {
-                  y(c,0,qx,qy,e) = u;
-                  y(c,1,qx,qy,e) = v;
+                  if (SDIM == 2)
+                  {
+                     double Jloc[4], Jinv[4];
+                     Jloc[0] = j(qx,qy,0,0,e);
+                     Jloc[1] = j(qx,qy,1,0,e);
+                     Jloc[2] = j(qx,qy,0,1,e);
+                     Jloc[3] = j(qx,qy,1,1,e);
+                     kernels::CalcInverse<2>(Jloc, Jinv);
+                     const double U = Jinv[0]*du[0] + Jinv[1]*du[1];
+                     const double V = Jinv[2]*du[0] + Jinv[3]*du[1];
+                     du[0] = U;
+                     du[1] = V;
+                  }
+                  else
+                  {
+                     double Jloc[6], Jinv[6];
+                     Jloc[0] = j(qx,qy,0,0,e);
+                     Jloc[1] = j(qx,qy,1,0,e);
+                     Jloc[2] = j(qx,qy,2,0,e);
+                     Jloc[3] = j(qx,qy,0,1,e);
+                     Jloc[4] = j(qx,qy,1,1,e);
+                     Jloc[5] = j(qx,qy,2,1,e);
+
+                     double ee, gg, ff;
+                     ee = Jloc[0]*Jloc[0] + Jloc[1]*Jloc[1] + Jloc[2]*Jloc[2];
+                     gg = Jloc[3]*Jloc[3] + Jloc[4]*Jloc[4] + Jloc[5]*Jloc[5];
+                     ff = Jloc[0]*Jloc[3] + Jloc[1]*Jloc[4] + Jloc[2]*Jloc[5];
+                     const double t = 1.0 / (ee*gg - ff*ff);
+                     ee *= t; gg *= t; ff *= t;
+
+                     Jinv[0] = Jloc[0]*gg - Jloc[3]*ff;
+                     Jinv[1] = Jloc[3]*ee - Jloc[0]*ff;
+                     Jinv[2] = Jloc[1]*gg - Jloc[4]*ff;
+                     Jinv[3] = Jloc[4]*ee - Jloc[1]*ff;
+                     Jinv[4] = Jloc[2]*gg - Jloc[5]*ff;
+                     Jinv[5] = Jloc[5]*ee - Jloc[2]*ff;
```

Comment:
Don't you think it makes sense to wrap this into a function in the `kernels` namespace?

Reply:
ping: @pazner

Reply:
Good idea. How does 6b5db0c503c8899930329332769f12011006ed4c look to you?

Reply:
I think this greatly improves the readability of the code for anyone who is not familiar with embedded elements!